/**
 * @file lifeindexwidget.cpp
 * @brief ç”Ÿæ´»æŒ‡æ•°å±•ç¤ºç»„ä»¶å®ç°
 */

#include "lifeindexwidget.h"
#include "ui_lifeindexwidget.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QDate>

LifeIndexWidget::LifeIndexWidget(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::LifeIndexWidget)
{
    ui->setupUi(this);
    setupConnections();
}

LifeIndexWidget::~LifeIndexWidget()
{
    delete ui;
}

void LifeIndexWidget::setupConnections()
{
    connect(ui->refreshBtn, &QPushButton::clicked,
            this, &LifeIndexWidget::onRefreshClicked);
}

void LifeIndexWidget::setCity(const QString &cityId, const QString &cityName)
{
    m_currentCityId = cityId;
    m_currentCityName = cityName;
    ui->cityLabel->setText(cityName);
    ui->dateLabel->setText(QDate::currentDate().toString("yyyyå¹´MMæœˆddæ—¥"));
    clear();
}

void LifeIndexWidget::updateLifeIndex(const QList<LifeIndex> &indices)
{
    clearIndexCards();
    
    int row = 0, col = 0;
    const int maxCols = 3;
    
    for (const LifeIndex &index : indices) {
        QFrame *card = createIndexCard(index);
        ui->indexGridLayout->addWidget(card, row, col);
        m_indexCards.append(card);
        
        col++;
        if (col >= maxCols) {
            col = 0;
            row++;
        }
    }
    
    // æ·»åŠ å¼¹æ€§ç©ºé—´
    ui->indexGridLayout->setRowStretch(row + 1, 1);
}

void LifeIndexWidget::clear()
{
    clearIndexCards();
}

void LifeIndexWidget::clearIndexCards()
{
    for (QFrame *card : m_indexCards) {
        ui->indexGridLayout->removeWidget(card);
        delete card;
    }
    m_indexCards.clear();
}

QFrame* LifeIndexWidget::createIndexCard(const LifeIndex &index)
{
    QFrame *card = new QFrame();
    card->setStyleSheet(
        "QFrame { background-color: white; border-radius: 12px; border: 1px solid #e4e7ed; }"
    );
    card->setMinimumHeight(140);
    
    QVBoxLayout *layout = new QVBoxLayout(card);
    layout->setSpacing(8);
    layout->setContentsMargins(16, 16, 16, 16);
    
    // é¡¶éƒ¨ï¼šå›¾æ ‡å’Œåç§°
    QHBoxLayout *headerLayout = new QHBoxLayout();
    
    QLabel *iconLabel = new QLabel(getIndexIcon(index.type));
    iconLabel->setStyleSheet("font-size: 28px;");
    headerLayout->addWidget(iconLabel);
    
    QLabel *nameLabel = new QLabel(index.name);
    nameLabel->setStyleSheet("font-size: 16px; font-weight: bold; color: #303133;");
    headerLayout->addWidget(nameLabel);
    
    headerLayout->addStretch();
    
    // ç­‰çº§æ ‡ç­¾
    QLabel *levelLabel = new QLabel(index.category);
    levelLabel->setStyleSheet(QString(
        "font-size: 12px; font-weight: bold; padding: 4px 8px; "
        "border-radius: 4px; background-color: %1; color: white;"
    ).arg(getLevelColor(index.level)));
    headerLayout->addWidget(levelLabel);
    
    layout->addLayout(headerLayout);
    
    // æè¿°
    QLabel *descLabel = new QLabel(index.description);
    descLabel->setWordWrap(true);
    descLabel->setStyleSheet("font-size: 13px; color: #606266; line-height: 1.5;");
    descLabel->setMinimumHeight(50);
    layout->addWidget(descLabel);
    
    layout->addStretch();
    
    return card;
}

void LifeIndexWidget::onRefreshClicked()
{
    if (!m_currentCityId.isEmpty()) {
        emit refreshRequested(m_currentCityId);
    }
}

QString LifeIndexWidget::getIndexIcon(const QString &type)
{
    static QMap<QString, QString> iconMap = {
        {"1", "ğŸƒ"},   // è¿åŠ¨æŒ‡æ•°
        {"2", "ğŸš—"},   // æ´—è½¦æŒ‡æ•°
        {"3", "ğŸ‘”"},   // ç©¿è¡£æŒ‡æ•°
        {"4", "ğŸ£"},   // é’“é±¼æŒ‡æ•°
        {"5", "â˜€ï¸"},   // ç´«å¤–çº¿æŒ‡æ•°
        {"6", "ğŸŒ¡ï¸"},  // æ—…æ¸¸æŒ‡æ•°
        {"7", "ğŸ¤§"},   // è¿‡æ•æŒ‡æ•°
        {"8", "ğŸ˜·"},   // èˆ’é€‚åº¦æŒ‡æ•°
        {"9", "ğŸ¤’"},   // æ„Ÿå†’æŒ‡æ•°
        {"10", "ğŸŒ¬ï¸"}, // ç©ºæ°”æ±¡æŸ“æ‰©æ•£æ¡ä»¶æŒ‡æ•°
        {"11", "ğŸŒ¤ï¸"}, // ç©ºè°ƒå¼€å¯æŒ‡æ•°
        {"12", "ğŸ•¶ï¸"}, // å¤ªé˜³é•œæŒ‡æ•°
        {"13", "ğŸ’„"},  // åŒ–å¦†æŒ‡æ•°
        {"14", "ğŸ§º"},  // æ™¾æ™’æŒ‡æ•°
        {"15", "ğŸšŒ"},  // äº¤é€šæŒ‡æ•°
        {"16", "ğŸ›¡ï¸"}  // é˜²æ™’æŒ‡æ•°
    };
    
    return iconMap.value(type, "ğŸ“Š");
}

QString LifeIndexWidget::getLevelColor(const QString &level)
{
    bool ok;
    int lvl = level.toInt(&ok);
    if (!ok) lvl = 3;
    
    if (lvl <= 1) return "#67C23A";      // ä¼˜/é€‚å®œ - ç»¿è‰²
    if (lvl <= 2) return "#409EFF";      // è‰¯/è¾ƒé€‚å®œ - è“è‰²
    if (lvl <= 3) return "#E6A23C";      // ä¸­ç­‰ - é»„è‰²
    if (lvl <= 4) return "#F56C6C";      // è¾ƒå·® - æ©™è‰²
    return "#909399";                     // å·® - ç°è‰²
}
